lib/open_project/translations/models/locales_updater.rb

```diff
@@ -0,0 +1,157 @@
+
+ENGLISH_TRANLATION_FILE = 'en.yml'
+ENGLISH_JS_TRANLATION_FILE = 'js-en.yml'
+ACCEPTANCE_LEVEL = ENV['ACCEPTANCE_LEVEL'] || 100
+
+class LocalesUpdater
+  extend TmpDirectory
+
+  def self.update_all_locales_of_all_repos(debug: true)
+    repos_to_update = plugins_with_locales
+
+    within_tmp_directory(delete_if_exists: true, debug: debug) do
+      repos_to_update.each do |name, specifics|
+        # todo each branch
+        create_i18n_handle(specifics)
+
+        within_tmp_directory(path: File.join(FileUtils.pwd, name), debug: debug) do
+          uri = specifics[:uri]
+          git_repo = GitRepository.new(uri, FileUtils.pwd)
+          git_repo.clone
+
+          git_repo.checkout(branch)
+          # todo or should we merge this branch into the next ('push vs pull')
+          git_repo.merge(previous_branch, strategy: :ours) if previous_branch
+          git_repo.within_repo do
+            # todo rescue should be in provider class
+            begin
+              upload_english
+
+              request_build
+              download_and_replace_locales
+            rescue Crowdin::API::Errors::Error => e
+              puts "Error during update of #{name}: #{e.message}"
+              exit
+            end
+          end
+          git_repo.add('config/locales')
+          git_repo.commit('update locales from crowdin')
+          git_repo.push(branch) unless debug
+        end
+      end
+    end
+  end
+
+  def self.plugins_with_locales
+    configuration[:plugins]
+  end
+
+  def self.configuration
+    LocalesUpdaterConfiguration.configuration
+  end
+
+  def self.branch
+    configuration[:branch]
+  end
+
+  def self.previous_branch
+    configuration[:previous_branch]
+  end
+
+  def self.create_i18n_handle(configuration_hash)
+    @i18n_provider ||= begin
+      project_id = configuration_hash[:project_id]
+      api_key = configuration_hash[:api_key]
+      version = configuration_hash[:version]
+      I18nProvider.new(project_id, api_key, version)
+    end
+  end
+
+  def self.upload_english
+    # either add or update the english (js) translation file
+    titles = {
+      ENGLISH_TRANLATION_FILE => 'OpenProject Wording',
+      ENGLISH_JS_TRANLATION_FILE => 'OpenProject JavaScript Wording'
+    }
+    [ENGLISH_TRANLATION_FILE, ENGLISH_JS_TRANLATION_FILE].each do |translation_file|
+      path_to_translation = File.join 'config', 'locales', translation_file
+      title = titles[translation_file]
+      @i18n_provider.upload_english(translation_file, path_to_translation, title)
+    end
+  end
+
+  def self.request_build
+    @i18n_provider.request_build
+  end
+
+  def self.download_and_replace_locales
+    # todo delete all locales here? maybe for the case that
+    # we do not support a language anymore.
+    begin
+      languages_files = create_temp_file('crowdin_translations')
+      @i18n_provider.download_locales(languages_files.path)
+
+      target_directory = Pathname(File.join('config', 'locales'))
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.